packages/xo-web/src/xo-app/sr/tab-general.js

```diff
@@ -32,9 +75,85 @@ export default class TabGeneral extends Component {
     }
   }
 
-  render() {
-    const { sr, vdis, vdiSnapshots, unmanagedVdis } = this.props
+  _getDiskGroups = createSelector(
+    () => this.props.vdis,
+    () => this.props.vdiSnapshots,
+    () => this.props.unmanagedVdis,
+    (vdis, vdiSnapshots, unmanagedVdis) => {
+      const disksParent = []
+      const groups = []
+      const snapshotsByVdis = groupBy(vdiSnapshots, '$snapshot_of')
+      const unmanagedVdisById = keyBy(unmanagedVdis, 'id')
+      let bc
+      let nBaseCopy = 0
+      let nSnapshots = 0
+      let snapshots
+      let usageBaseCopies = 0
+      let usageSnapshots = 0
+      if ((snapshots = snapshotsByVdis[undefined]) !== undefined) {
+        snapshots.forEach(snapshot => {
+          groups.push(snapshot)
+        })
+      }
+      forEach(vdis, vdi => {
+        const { id, parent } = vdi
+        if (parent !== undefined) {
+          bc = unmanagedVdisById[parent]
+          while (bc !== undefined) {
+            nBaseCopy += 1
+            usageBaseCopies += bc.usage
+            disksParent.push(bc)
+            bc =
+              bc.parent === undefined ? undefined : unmanagedVdisById[bc.parent]
+          }
+        }
+        if ((snapshots = snapshotsByVdis[id]) !== undefined) {
+          nSnapshots = snapshots.length
+          usageSnapshots = sumBy(snapshots, 'usage')
+        }
+        groups.push({
+          baseCopies:
+            nBaseCopy > 0
+              ? {
+                  n: nBaseCopy,
+                  usage: usageBaseCopies,
+                }
+              : undefined,
+          id,
+          n: nBaseCopy + nSnapshots + 1,
+          snapshots:
+            nSnapshots > 0
+              ? {
+                  n: nSnapshots,
+                  usage: usageSnapshots,
+                }
+              : undefined,
+          usage: vdi.usage + usageBaseCopies + usageSnapshots,
+          vdi,
+        })
+        nBaseCopy = 0
+        nSnapshots = 0
+        usageBaseCopies = 0
+        usageSnapshots = 0
+        snapshots = undefined
+      })
+
+      const diff = differenceBy(unmanagedVdis, disksParent, 'id')
```

Comment:
When you're inside such a complex piece of code, try to find a better name than `diff`.